sw/device/lib/dif/dif_flash_ctrl.h

```diff
@@ -0,0 +1,1019 @@
+ * @return `kDifFlashCtrlBadArg` if `flash_ctrl` is null or if the value `rd` is
+ * out of range, `kDifFlashCtrlOk` otherwise.
+ */
+DIF_WARN_UNUSED_RESULT
+dif_flash_ctrl_result_t dif_flash_ctrl_fifo_prog_lvl_set(
+    const dif_flash_ctrl_t *flash_ctrl, uint32_t prog);
+
+/**
+ * Set the interrupt watermarks for the read FIFO.
+ *
+ * The value of `rd` defines the level the read FIFO must fill to before
+ * triggering a `rd_lvl` interrupt.
+ *
+ * This interrupt will only trigger if enabled through the interrurpt API.
+ *
+ * @param flash_ctrl flash controller to set FIFO level watermarks for.
+ * @param rd Trigger an interrupt when the read FIFO fills to this level.
+ * @return `kDifFlashCtrlBadArg` if `flash_ctrl` is null or if the value of
+ * `prog` is out of range, `kDifFlashCtrlOk` otherwise.
+ */
+DIF_WARN_UNUSED_RESULT
+dif_flash_ctrl_result_t dif_flash_ctrl_fifo_rd_lvl_set(
+    const dif_flash_ctrl_t *flash_ctrl, uint32_t rd);
+
+/**
+ * Get the interrupt watermarks for the program and read FIFOs.
+ *
+ * @sa #dif_flash_ctrl_fifo_lvl_set()
+ *
+ * @param flash_ctrl flash controller to get FIFO level watermarks from.
+ * @param prog_out Out-parameter, trigger an interrupt when the program FIFO
+ * drains to this level. May be null.
+ * @param rd_out Out-parameter, trigger an interrupt when the read FIFO fills to
+ * this level. May be null.
+ * @return `kDifFlashCtrlBadArg` if `flash_ctrl` is null, `kDifFlashCtrlOk`
+ * otherwise.
+ */
+DIF_WARN_UNUSED_RESULT
+dif_flash_ctrl_result_t dif_flash_ctrl_fifo_lvl_get(
+    const dif_flash_ctrl_t *flash_ctrl, uint32_t *prog_out, uint32_t *rd_out);
+
+/**
+ * Resets both the program and read FIFOs.
+ *
+ * This is useful in the event of an unexpected error as a means of reseting
+ * state.
+ *
+ * @param flash_ctrl flash controller device to clear FIFOs on.
+ * @return `kDifFlashCtrlBadArg` if `flash_ctrl` is null, `kDifFlashCtrlOk`
+ * otherwise.
+ */
+DIF_WARN_UNUSED_RESULT
+dif_flash_ctrl_result_t dif_flash_ctrl_fifo_reset(
+    const dif_flash_ctrl_t *flash_ctrl);
+
+/**
+ * Push data to the program FIFO.
+ *
+ * Attempts to write the contents of `data` to the program FIFO. It is required
+ * that a program transaction be started prior to calling this function, else
+ * the call will fail with `kDifFlashCtrlProgInactive`.
+ *
+ * The following conditions are also required:
+ *   - `data` must reference a contiguous, allocated, readable region of at
+ *     least `word_count` words, violation of this will produce undefined
+ *     behavior.
+ *   - The first call to this function after starting the program transaction
+ *     the same `word_count` must be used as was supplied at the start of the
+ *     program transaction.
+ *   - Each subsequent call the new `word_count` must equal `word_count -
+ *     words_sent_out` from the previous call.
+ * All deviations on the above will produce a `kDifFlashCtrlProgBadArg` error
+ * unless otherwise specified.
```

Comment:
Thanks for documenting this, but I'm not sure why we really need this. I understand that clients can always start multiple smaller transactions but I find it more intuitive to let clients start a transaction and call `dif_flash_ctrl_prog_fifo_push` until all words are written.